### ISSUE TITLE
could the security reports be described in terms of openssl APIs?

### ISSUE THREAD

**Issue by sam-github** [openssl/openssl#10622]
I've seen this a few times, so would like to suggest that if it is possible, the openssl security advisories would be more useful to users of openssl, if the security issues were described in terms of the openssl API.

Its not so easy for a non-cryptographer, or a non-expert in openssl internals to understand how to know if their use of openssl is vulnerable.

If the OpenSSL vuln advisories could be phrased in such a way that non-experts could audit their code to detemine applicability or non applicability, that would be very helpful.

For example:

> Attacks against DH512 are considered just feasible.

Does this mean only users of specific TLS cipher suites? Or of specific cipher suites, with openssl dh params generated in certain ways?

> However, for an attack the target would have to re-use the DH512 private key, which is not
recommended anyway.

Is reuse of the DH512 private key possible for a user of the TLS API?

> Also applications directly using the low level API BN_mod_exp may be affected if they use BN_FLG_CONSTTIME.

This is closer to useful.  It refers to a specific openssl API name and flag, so its possible to audit for this usage. The "may" takes away a bit from the usefulness. What other conditions are necessary to change the may into either a "not affected" or "is affected"?

> If an application already calls OPENSSL_init_crypto() explicitly using OPENSSL_INIT_ATFORK then this problem does not occur at all.

That is an excellent description, easy to search for.

>  In order for this to be exploitable "non-stitched" ciphersuites must be in use. Stitched ciphersuites are optimised implementations of certain commonly used ciphersuites.

This was a not so useful description. How is a user of OpenSSL APIs supposed to know if they use "optimised implementations of certain commonly used ciphersuites"?

**Comment by mspncp** [comment]
@sam-github sorry that this ticket was forgotton somehow. Your proposal actually deserves some attention by the @openssl/otc.

**Comment by nhorman** [comment]
@sam-github I'm sorry, but the action you seem to be requesting is somewhat non-actionable.  "Make security vulnerabilities easier to understand" is a good philosophical statement of goal, but its very subjective.  The examples you give are reasonable questions to ask, but security vulnerabilities assume a certain level of understanding of (a) the subject matter (b) the software that they target, so it becomes very difficult to write said vulnerabilities for unfamiliar parties without turning them effectively turning them into semi-complete manuals on the software in question

That said, specific improvements could be made, perhaps by creating a style guide for use when writing them.  Some examples:
* vulnerabilities for a TLS cipher suite should enumerate all ciphers affected
* If a term of art is used, the definition of said term should be stated
These are just examples of course.  Can you provide a more complete list of items to include in a vulnerability that would make them easier for you?

**Comment by sam-github** [comment]
>  "Make security vulnerabilities easier to understand" is a good philosophical statement of goal, but its very subjective.

My suggestion was quite specific, that recognizable names of constants/APIs/etc. from the OpenSSL API be used in announcements, so that code maintainers can easily *determine whether their code bases are affected, or not*. Maintainers don't need to "understand" the vulnerability, we need to know if applies to their code bases. 

I provided a list in the description of examples where the terms used have no obvious relationship to the OpenSSL API, and as such, are not particularly useful to maintainers, since there is not obvious path to figuring out if they do or do not apply to our code bases.

I also listed some examples where they were more useful, it wasn't all negative.

Describing vulnerabilities using terms of art from cryptography and mathematics is only useful to people who are familiar with the internals of OpenSSL, make some reports not so useful for *users* of OpenSSL.
